akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala

```diff
@@ -204,6 +204,17 @@ object Source {
     })
 
   /**
+   * Create [[Source]] that will continually produce given elements in specified order.
+   *
+   * Start a new 'cycled' `Source` from the given elements. The producer stream of elements
+   * will continue infinitely by repeating the same sequence of elements provided during construction.
+   */
+  def cycle[T](f: () ⇒ Iterator[T]): Source[T, NotUsed] = fromIterator(() ⇒ {
+    if (f().isEmpty) Iterator.empty
```

Comment:
Will do the update on docs regarding empty iterator case. Should I update `fromIterator` docs for the same case?

I would do the following:
```
  def cycle[T](f: () ⇒ Iterator[T]): Source[T, NotUsed] = {
    val iterator = f()
    val cycledArg = if (iterator.nonEmpty) Iterator.continually(iterator).flatten else iterator
    fromIterator(() ⇒ cycledArg).withAttributes(DefaultAttributes.cycledSource)
  }
```
But unfortunately this results infinite loop in scala internals when `iterator` runs out of elements.

So it looks like that the only way to 'cycle' iterator is to do it with repeated evaluation of `f`: `Iterator.continually(f()).flatten`.

Now regarding the empty case. Again you cannot provide empty iterator to `Iterator.continually` method (see [stack overflow question](http://stackoverflow.com/questions/36818050/iterator-continually-cannot-evaluate-expression-in-scala-repl)), it looks like that we need to do 'flattening' by ourselves returning empty iterator or `f()`: `if (f().isEmpty) Iterator.empty` or `vol iterator = f() if(iterator.isEmpty) iterator else Iterator.continually(f()).flatten`.

All said above leads me to the following implementation:
```
  def cycle[T](f: () ⇒ Iterator[T]): Source[T, NotUsed] = {
    val iterator = if (f().isEmpty) Iterator.empty else Iterator.continually(f()).flatten
    fromIterator(() ⇒ iterator).withAttributes(DefaultAttributes.cycledSource)
  }
```
Please let me know if you agree with it